### DISCUSSION TITLE
What would be the Easiest ways to distribute PyInfra?

### DISCUSSION THREAD

**Discussion by v3ss0n** [pyinfra-dev/pyinfra#955] [General]
I am trying to build pyinfra as a standalone build using nuitka. Had anyone tried it before?
The build was fine but when running , i am having Key not found ssh error   it seems the problem with Paramiko import.

I had tried like this :

`git clone https://github.com/Fizzadar/pyinfra`
`nuitka3 --follow-imports --standalone pyinfra/pyinfra_cli/main.py`

It builds the binary successfully but gives this errror when running  i get :

```python
File "<frozen importlib._bootstrap>", line 1027, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1006, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 688, in _load_unlocked
  File "/mnt/development/Workspace/SnG/pyinfra.repo/pyinfra/api/host.py", line 79, in <module pyinfra.api.host>
    class Host:
  File "/mnt/development/Workspace/SnG/pyinfra.repo/pyinfra/api/host.py", line 120, in Host
    executor=get_execution_connector("ssh"),
  File "/mnt/development/Workspace/SnG/pyinfra.repo/pyinfra/api/connectors.py", line 46, in get_execution_connector
    return get_execution_connectors()[name]
KeyError: 'ssh'
````
Had anyone tried this route? Maybe pyinfra/pyinfra_cli/main.py is not a good entry point.

**Comment by Fizzadar** [comment]
I have nearly got this working using `pyinstaller` in https://github.com/Fizzadar/pyinfra/pull/768, but it's painfully slow to start (affects all pyinstaller single file binaries, not pyinfra specific). I did also try in Nuitka but the resulting binary was just as slow.

The key error your seeing is because of the [hidden imports that are setuptools entrypoints](https://github.com/Fizzadar/pyinfra/blob/2.x/setup.py#L123). I think this may just mean you need to explicitly ensure `pyinfra.connectors` is imported?

**Reply by v3ss0n** [reply]
i saw how you did at https://github.com/Fizzadar/pyinfra/compare/2.x...nuitka-binaries  but i gave up and tried with AppImage , it works and much easier to build , so fast to build too.
Update : i sent PR for PyInfra AppImage builder : https://github.com/Fizzadar/pyinfra/pull/957

**Reply by v3ss0n** [reply]
What do you think about appimage that I contributed?
